### DISCUSSION TITLE
Generating a localized URL to the current Rodauth page

### DISCUSSION THREAD

**Discussion by janko** [jeremyevans/rodauth#380] [General]
I'm re-implementing rubygems.org authentication using Rodauth for my own interest. This app has a language switcher which is a link to the current page with the locale added as a query parameter. Another option for localizing routes is using locale prefixes, as I showed in an example for [rodauth-i18n](https://github.com/janko/rodauth-i18n?tab=readme-ov-file#localized-routes).

I would like to find an easy way to implement such a language switcher, ideally something that would work for both localization strategies. This is easy to do when I know the action I'm generating the URL for:

```rb
rodauth.login_path(locale: "hr")
# or
I18n.with_locale("hr") { rodauth.login_path }
```

However, if I want to add this to the layout template, I don't a reliable way to determine what Rodauth route am I currently on. I was considering matching the current path with `Rodauth::Auth.route_hash`, but that seemed cumbersome and not 100% bullet proof.

I was thinking that ideally Rodauth would provide a `#current_path` and/or `#current_url` method, that would delegate to the corresponding `*_path` or `*_url` route helper. Alternatively, maybe Rodauth could set the current action in an instance variable?

```diff
diff --git a/lib/rodauth.rb b/lib/rodauth.rb
index 8a93a0c..5ee8867 100644
--- a/lib/rodauth.rb
+++ b/lib/rodauth.rb
@@ -139,6 +139,7 @@ module Rodauth
       define_method(handle_meth) do
         request.is send(route_meth) do
           check_csrf if check_csrf?
+          @action = name
           _around_rodauth do
             before_rodauth
             send(internal_handle_meth, request)
```

**Comment by jeremyevans** [comment]
`rodauth.request.path` should work to return the path (this comes from Rack).  Do you think that would work for you?

**Reply by janko** [reply]
This would require re-implementing the localization in the URL. For example, if I'm on `/en/login` and want to switch to `hr` locale, then I'd need to parse the current path to replace the `en` with `hr`. There might not even be a locale prefix in the URL e.g. for the default english locale. Query parameters would also ideally need to be preserved, which is a bit cumbersome in Rack. For example:

```rb
localized_path = request.path.sub(/^\/#{I18n.locale}/, locale)
localized_path += "?#{Rack::Utils.buils_query(request.query_parameters)}" if request.query_parameters.any?
```

It's definitely possible, it just requires thought and knowledge of Rack. In Rails this is just `url_for(locale: "hr")` for declared routes, the developer doesn't need to think how to construct the URL. I was wondering if it could be made similarly easy with Rodauth, e.g. by finding a way to use Rodauth's `*_path` and `*_url` helpers.

**Reply by jeremyevans** [reply]
I'm definitely open to making what you want easier, as long as the approach is backwards compatible.  Setting an instance variable as part of route dispatch would be OK as part of the solution.
